src/golang/cmd/server/server/aqueduct_server.go

```diff
@@ -298,3 +310,21 @@ func IndexHandler() func(w http.ResponseWriter, r *http.Request) {
 
 	return http.HandlerFunc(fn)
 }
+
+// Pause puts the server in system maintenance mode by blocking all new requests
+// and waits for all active requests to finish.
+// It is the responsibility of the caller to call s.Restart() to allow requests
+// to be processed again once the system maintenance is complete.
+func (s *AqServer) Pause() {
+	s.UnderMaintenance.Store(true)
+	s.RequestMutex.Lock()
+}
+
+// Restart restarts a server that was previously stopped via s.Pause().
+func (s *AqServer) Restart() {
+	if err := s.Init(); err != nil {
+		log.Fatalf("Unable to restart server: %v", err)
+	}
+	s.RequestMutex.Unlock()
```

Comment:
why do we unlock before setting maintenance mode to false?

Reply:
We're just doing it in the opposite order that they were acquired. As soon as we set maintenance mode to false, new requests will start coming in, but until we unlock the `RequestMutex` they won't be able to proceed.